### PR TITLE
initial

### PR DESCRIPTION
**Pull request by a10zn8**

### CODE REVIEW

File: run.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	SourceHost string `long:"source" short:"s" description:"source eth host" default:"https://eth.drpc.org"`
+	Scenario   string `long:"scenario" short:"c" description:"scenario file" default:"stages.json"`
+	Result     string `long:"result" short:"r" description:"result file" default:"result.json"`
+}
+
+func main() {
+	options := Options{}
+	_, err := flags.Parse(&options)
+	if err != nil {
+		return
```
Comment (KirillPamPam) [inline run.go:29]: Maybe also add `fmt.Println(err)`?

File: run.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+	}
+	fmt.Println("Start")
+
+	jsonString, err := os.ReadFile(options.Scenario)
+	if err != nil {
+		fmt.Println(err)
+	}
```
Comment (KirillPamPam) [inline run.go:36]: Add `return` to stop executing?

File: run.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+	var stages []Stage
+	err = json.Unmarshal(jsonString, &stages)
+	if err != nil {
+		fmt.Println(err)
+	}
```
Comment (KirillPamPam) [inline run.go:42]: Also `return`?

File: run.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+			attacker := vegeta.NewAttacker()
+
+			var metrics vegeta.Metrics
+			for res := range attacker.Attack(targeter, rate, duration, "Big Bang!") {
+				metrics.Add(res)
+				stageMetrics.Add(res)
+			}
+			attacker.Stop()
+			metrics.Close()
+			cancel()
```
Comment (KirillPamPam) [inline run.go:72]: I'm not sure, but is it better to use `defer cancel()` before all this code?

File: run.go (hunk 1 above)
Comment (a10zn8) [reply run.go:29]: flags.Parse also prints error

File: run.go (hunk 4 above)
Comment (a10zn8) [reply run.go:72]: defer executes at the end of function scope, but here i need to close resources on each iteration of loop

**Review by KirillPamPam — APPROVED** [review]

File: run.go [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+		Client: client,
+	}
+
+	stateChannel := make(chan ethspam.State, 1)
+
+	randSrc := rand.NewSource(time.Now().UnixNano())
+	go func() {
+		state := ethspam.LiveState{
+			IdGen:   &ethspam.IdGenerator{},
+			RandSrc: randSrc,
```
Comment (scripter-v) [inline run.go:122]: `rand.Source` isn't thread safe so it's better to make it inside goroutine.

File: run.go [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+			case <-ctx.Done():
+				return
+			default:
+			}
+			if q, err := generator.Query(state); err == io.EOF {
+				return
+			} else if err != nil {
+				exit(2, "failed to write generated query: %s", err)
+			} else {
+				queries <- q.GetBody()
```
Comment (scripter-v) [inline run.go]: We can hang here, and this will lead goroutines leak. The best solution is to add another select here.

File: run.go [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+			}
+		}
+	}()
+
+	state := <-stateChannel
+
+	queries := make(chan string)
+
+	go func() {
+		for {
```
Comment (scripter-v) [inline run.go:163]: Let's add `defer close(queries)` before loop

File: run.go [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+	stateChannel := make(chan ethspam.State, 1)
+
+	randSrc := rand.NewSource(time.Now().UnixNano())
+	go func() {
+		state := ethspam.LiveState{
+			IdGen:   &ethspam.IdGenerator{},
+			RandSrc: randSrc,
+		}
+		for {
```
Comment (scripter-v) [inline run.go:125]: Let's add `defer close(stateChannel)` before loop

File: run.go [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+		stagesResults = append(stagesResults, stageResult)
+		fmt.Printf("Stage %v - succ: %v - mean: %v - p90: %v \n", stage.Name, stageMetrics.Success, stageMetrics.Latencies.Mean, stageMetrics.Latencies.P90)
+	}
+	fmt.Println("Finish")
+
+	res, _ := json.Marshal(stagesResults)
+	os.WriteFile(options.Result, res, 0644)
+}
+
+func NewEthSpamTargeter(host string, queryParams map[string]int64, parentHost string, ctx context.Context) vegeta.Targeter {
```
Comment (scripter-v) [inline run.go]: context is usually put first

File: run.go [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+			select {
+			case stateChannel <- newState:
+			case <-ctx.Done():
+				return
+			}
+
+			select {
+			case <-time.After(15 * time.Second):
+			case <-ctx.Done():
+			}
```
Comment (scripter-v) [inline run.go:153]: should be `return` in `ctx.Done` case

**Review by scripter-v — CHANGES_REQUESTED** [review]

**Review by scripter-v — APPROVED** [review]

**Pull request merged**